Allow constant integer variables as array lengths. Closes #716

Balajiganapathi opened the pull request:
Fixes #716.

I have extended ConstantEvaluator to support identifiers. It also recursively calls in case of more constant references.

Balajiganapathi committed:
Allow constant integer variables as array lengths.

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+{
+	VariableDeclaration const *variableDeclaration = dynamic_cast<VariableDeclaration const *>(_identifier.annotation().referencedDeclaration);
+	if (!variableDeclaration)
+		return;
```
There should be a test for this, ie. `function f() {} uint[f] tmp;`

axic replied:
Test was added.

axic commented:
Thanks for this! There should be a lot more tests because this a new feature. You must also ensure the case about pure expressions in constants, such as `uint constant x = keccak256(ripemd160(33))`, which is really tricky because the `ripemd160` part is a `CALL` and unfortunately it is not evaluated once at assignment, but rather every time it is used...

Balajiganapathi committed:
Add more tests for constant var as array lengths

Balajiganapathi commented:
@axic I have added more tests. 

> You must also ensure the case about pure expressions in constants, such as uint constant x = keccak256(ripemd160(33))

Do you mean we should allow such expressions? In current implementation they are not allowed ( I have added a test for that case too).

axic commented:
> > You must also ensure the case about pure expressions in constants, such as uint constant x = keccak256(ripemd160(33))
> Do you mean we should allow such expressions?

I'd keep disallowing them so that's great the way it is.

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+
+void ConstantEvaluator::endVisit(Identifier const& _identifier)
+{
+	VariableDeclaration const *variableDeclaration = dynamic_cast<VariableDeclaration const *>(_identifier.annotation().referencedDeclaration);
```
Style: `const* variableDeclaration` and `<VariableDeclaration const*>`

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+	else
+	{
+		ASTPointer<Expression> value = variableDeclaration->value();
+		if (value)
```
In what case doesn't it have a value and still counts as valid for array length?

Balajiganapathi replied:
Oh, it won't be valid in that case. WIll throw an error in else clause.

Balajiganapathi replied:
Done

chriseth commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+		if (value)
+		{
+			if (!value->annotation().type)
+				ConstantEvaluator e(*value, m_errorReporter);
```
Are we sure this cannot lead to infinite recursion at this point in the compiler? Also, is the recursion depth limited?

chriseth replied:
For the second point, it is fine to just add a counter in ConstantEvaluator and fold if it gets too high.

Balajiganapathi replied:
@chriseth Can't think of a scenario that can lead to infinite recursion. It is a good idea to have a recursion depth limit though. What depth would be a good one? 10? 20? How do we decide? And what error to throw in that case?

Balajiganapathi replied:
@chriseth You are correct, a simple "uint constant LEN = LEN;" caused an infinite recursion (and crashed my laptop :( ). Will add a recursion limit of 20 (?).

Balajiganapathi replied:
Done.

Balajiganapathi committed:
Detect cyclic constant definitions

axic commented:
> Will add a recursion limit of 20 (?).

I like powers of two, perhaps 32?

Please include tests for recursion: `uint constant LEN = LEN;` and `uint constant A = B; uint constant B = A;`

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -7221,6 +7221,151 @@ BOOST_AUTO_TEST_CASE(array_length_not_convertible_to_integer)
+{
+	char const* text = R"(
+		contract C {
+            uint constant LEN = LEN;
```
Indentation is off here.

Balajiganapathi replied:
Weird. This looks fine on my ide. Will try to find out why this is happening.

axic commented on test/libsolidity/SolidityNameAndTypeResolution.cpp:
```diff
@@ -7221,6 +7221,151 @@ BOOST_AUTO_TEST_CASE(array_length_not_convertible_to_integer)
+{
+	char const* text = R"(
+		contract C {
+            uint constant L2 = LEN - 10;
```
Indentation is off here.

axic commented on libsolidity/analysis/ConstantEvaluator.h:
```diff
@@ -38,8 +38,9 @@ class TypeChecker;
 public:
-	ConstantEvaluator(Expression const& _expr, ErrorReporter& _errorReporter):
-		m_errorReporter(_errorReporter)
+	ConstantEvaluator(Expression const& _expr, ErrorReporter& _errorReporter, size_t _newDepth):
```
Could have default value for `_newDepth = 0` so that `ReferencesResolver` doesn't need to be updated.

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,34 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+		return;
+	if (!variableDeclaration->isConstant())
+	{
+		m_errorReporter.fatalTypeError(_identifier.location(), "Identifier must be declared constant.");
```
We consider `fatalTypeError` to be throwing so it doesn't need to have an else branch.

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,34 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+		}
+		else
+		{
+			m_errorReporter.fatalTypeError(_identifier.location(), "Constant identifier declaration must have a constant value.");
```
Same applies here as above, so I'd reorder this as `if (!value) fatalTypeError`.

axic commented on libsolidity/analysis/ConstantEvaluator.h:
```diff
@@ -48,8 +49,11 @@ class ConstantEvaluator: private ASTConstVisitor
 
 	ErrorReporter& m_errorReporter;
+	size_t depth; // The recursion depth; to avoid infinite recursion caused by cyclic constant definition
+	size_t DEPTH_LIMIT = 32;
```
If this constant is kept here it should be `static const` and probably named `RECURSION_DEPTH_LIMIT`. Though looking at the other PR consensus seems to be including the value in the code instead.

axic commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,34 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+void ConstantEvaluator::endVisit(Identifier const& _identifier)
+{
+	if (depth > DEPTH_LIMIT)
+	{
```
As per the "style" parentheses are not required for single statement bodies.

axic commented:
I think that's all the comments, should be done now :)

Balajiganapathi commented:
Done @axic :)

axic committed:
Improve cyclic constant error message

chriseth commented on libsolidity/analysis/ConstantEvaluator.cpp:
```diff
@@ -74,3 +74,25 @@ void ConstantEvaluator::endVisit(Literal const& _literal)
+	if (!variableDeclaration)
+		return;
+	if (!variableDeclaration->isConstant())
+		m_errorReporter.fatalTypeError(_identifier.location(), "Identifier must be declared constant.");
```
Shouldn't this and the next be assertions?
